ambry-router/src/main/java/com/github/ambry/router/PutOperation.java

```diff
@@ -574,10 +575,23 @@ void setOperationCompleted() {
 
   /**
    * Clean up the chunks to release any data buffer. This should be invoked when terminating the operation with
-   * an exception.
+   * an exception. This method also closes the chunkFillerChannel to fire any pending callbacks, ensuring the original
+   * buffer from the ReadableStreamChannel is properly released. Synchronized for memory visibility on channelReadBuf.
+   * The contract upheld by PutManager is that this method is called AT-MOST-ONCE.
    */
-  public void cleanupChunks() {
-    releaseDataForAllChunks();
+  public synchronized void cleanupChunks() {
+    try {
+      releaseDataForAllChunks();
+    } finally {
+      // Release the extra reference we retained when storing in channelReadBuf.
+      if (channelReadBuf != null) {
+        channelReadBuf.release();
+        channelReadBuf = null;
+      }
+      // Close the chunkFillerChannel to fire any remaining callbacks in chunksAwaitingResolution.
+      // This ensures the original buffer (owned by the callback) is released and not leaked.
+      chunkFillerChannel.close();
```

Comment:
There are two cases we need to close the chunkFillerChannel to clean up the buffered chunks. 
1. Client close the connection before sending all the bytes
2. PutOperation failed to write to ambry servers.

In both cases, an exception would be send back to the `NettyResponseChannel` to close `NettyRequest`. When `NettyRequest` is closed, it would call the callback method we passed to int `readInto` method, which would close the chunk filler channel. 

`ByteBufferAsyncWritableChannel` has an atomic boolean variable to make sure the channel is closed only once, so calling `close` twice won't do any harm. But it's also not necessary.

Reply:
Moving `cleanupChunks` to readnto callback is not a bad idea. I would prefer it that way. This way, we can make it a private method and don't have to call it from `PutManager`. It's probably better in this way.